js/initcontext.go

```diff
@@ -288,13 +71,68 @@ func readFile(fileSystem afero.Fs, filename string) (data []byte, err error) {
 }
 
 // allowOnlyOpenedFiles enables seen only files
-func (i *InitContext) allowOnlyOpenedFiles() {
-	fs := i.filesystems["file"]
-
+func allowOnlyOpenedFiles(fs afero.Fs) {
 	alreadyOpenedFS, ok := fs.(fsext.OnlyCachedEnabler)
 	if !ok {
 		return
 	}
 
 	alreadyOpenedFS.AllowOnlyCached()
 }
+
+type requireImpl struct {
+	vu      modules.VU
+	modules *moduleSystem
+	pwd     *url.URL
+}
+
+func (r *requireImpl) require(specifier string) (*goja.Object, error) {
+	// TODO remove this in the future when we address https://github.com/grafana/k6/issues/2674
+	// This is currently needed as each time require is called we need to record it's new pwd
+	// to be used if a require *or* open is used within the file as they are relative to the
+	// latest call to require.
+	// This is *not* the actual require behaviour defined in commonJS as it is actually always relative
+	// to the file it is in. This is unlikely to be an issue but this code is here to keep backwards
+	// compatibility *for now*.
+	// With native ESM this won't even be possible as `require` might not be called - instead an import
+	// might be used in which case we won't be able to be doing this hack. In that case we either will
+	// need some goja specific helper or to use stack traces as goja_nodejs does.
+	currentPWD := r.pwd
+	if specifier != "k6" && !strings.HasPrefix(specifier, "k6/") {
+		defer func() {
+			r.pwd = currentPWD
+		}()
+		// In theory we can give that downwards, but this makes the code more tightly coupled
+		// plus as explained above this will be removed in the future so the code reflects more
+		// closely what will be needed then
+		fileURL, err := loader.Resolve(r.pwd, specifier)
+		if err != nil {
+			return nil, err
+		}
+		r.pwd = loader.Dir(fileURL)
```

Comment:
Help me understand this hack... :sweat_smile: I read through and understand #2674, just not why we're doing _this_.

So you store the current `requireImpl.pwd` so you can revert it in the `defer`, but why do we need to change it here? The `return` below calls `r.modules.Require()` with `currentPWD`, so it doesn't seem the new WD is used(?).

In general, I'm not clear why we have a global reference to a working directory at all that needs to be changed. Couldn't each imported module have its own `pwd`, and we can resolve the paths to modules they import using the parent module's `pwd`? :confused: This way we could avoid keeping and reverting global state.

But I don't have the full picture here, so would appreciate some explanation.

Reply:
> Couldn't each imported module have its own pwd

No because that will be the breaking change that I describe in #2674 ;) 

Or at least we could, and I would argue we *should*, but not in this PR.

This code as well as practically all the module resolving and loading is just broken down the previous code. 

You can see more or less this exact code in the previous version:
https://github.com/grafana/k6/blob/b750e206859a495b8a1fe12aa016dc8ed0e1c9db/js/initcontext.go#L186-L187

This code specifically does one thing and that is to make the `open` and `require` relative to the file that is currently being evaluated within `require`. Which is *different* from the file it is in which is basically what you are proposing and how `require` is supposed to work under commonJS.

How this currently works depend on where it is called from and when and not on just where it is written, Example:

file `/path/to/B/B.js`: 
```javascript
exports.load = func(s) {
  //do somethign cool 
  require(s); // by commonjs standard this is *always* relative to /path/to/B/B.js
  // but in our case it is relative to whatever 
  // do somethign else cool
}
```
we also have file `/path/to/C/C.js` which contains are not relevant

file /path/to/A.js has
```javascript
let s = require("./B/B.js"); // I will be using `require` instead of ESM for the purpose of being simpler ;) 
s.load("./C/C.js"); // by the standard this will require `/path/to/B/C/C.js` but k6 has always inturpretted this as `/path/to/C/C.js`
```

This  is likely more heavily used with `open` where you might want to do something in particular. 

So in order for this to work `require` (and `open`) need to  figure out which is the currently being "required" module, and is there could be cycles, and 1 module can be required multiple times I don't think there is another way to do it that isn't stack based which is how I would propose we do it in a future PR. 

If you want to try at it or have a specific proposal - I am all ears. But this is kept this way specifically for this to not be  a breaking change and so it does have minimal chance of breaking anything - as it works more or less exactly as before.

I do expect another problem is that ... the way this works in k6 is *confusing* and almost the exact way you expect it to work or should work by how the specification is written. 

Functions similar to `open` for example seem to always work relative to the *root* script/module and not based on the file it is called in.

Also with connected to https://github.com/grafana/k6/pull/2881#discussion_r1095753228 : this is the code of the call of `require` but this change to the pwd is for the `require` or `open` calls *after* this one. This `require` call uses the `currentPWD` *specifically* so that it can know what *it* is relative to.

Reply:
I see, thanks for explaining.

I see now that this is what we were doing before, so agree with not changing the behavior in this PR.

I don't have a specific proposal in mind, so let's keep this as is.